File: platforms/ios/Packages/FunputKit/Sources/KeyboardLayout/Alternates/KeyAlternate.swift
import Foundation

public struct KeyAlternate: Hashable, Sendable, Identifiable {
    public let text: String
    public let shiftedText: String
    public let accessibilityLabel: String

    public var id: String { text }

    public init(
        text: String,
        shiftedText: String? = nil,
        accessibilityLabel: String? = nil
    ) {
        precondition(!text.isEmpty, "Alternate text must not be empty")
        self.text = text
        self.shiftedText = shiftedText ?? text.uppercased()
        self.accessibilityLabel = accessibilityLabel ?? text
    }

    public func text(for shiftState: ShiftState) -> String {
        shiftState.isUppercase ? shiftedText : text
    }
}

File: platforms/android/keyboard-renderer/src/main/java/app/funput/funput/keyboard/popover/model/KeyAlternate.kt
package app.funput.funput.keyboard.popover.model

import app.funput.funput.keyboard.model.ShiftState

sealed interface KeyAlternate {
    val accessibilityLabel: String

    data class Text(
        val text: String,
        val shiftedText: String = text.uppercase(),
        override val accessibilityLabel: String = text,
    ) : KeyAlternate {
        init {
            require(text.isNotEmpty()) { "Alternate text must not be empty" }
            require(shiftedText.isNotEmpty()) { "Shifted alternate text must not be empty" }
            require(accessibilityLabel.isNotBlank()) { "Alternate accessibility label must not be blank" }
        }

        fun textFor(shiftState: ShiftState): String =
            if (shiftState.isActive) shiftedText else text
    }

    enum class Action(override val accessibilityLabel: String) : KeyAlternate {
        PLACEMENT("Tùy chỉnh vị trí bàn phím"),
        SETTINGS("Cài đặt, mở ứng dụng Funput"),
    }
}
